sdk/go/property/path.go

```diff
@@ -0,0 +1,215 @@
+		v, err = segment.apply(v)
+		if err != nil {
+			return Value{}, err
+		}
+	}
+	return v, nil
+}
+
+// Set a value.
+//
+// Set does not change src in place, rather producing a copy.
+func (p Path) Set(src, to Value) (Value, PathApplyFailure) {
+	if len(p) == 0 {
+		return to, nil
+	}
+	butLast, last := p[:len(p)-1], p[len(p)-1]
+	v, err := butLast.Get(src)
+	if err != nil {
+		return Value{}, err
+	}
+	switch {
+	case v.IsArray():
+		i, ok := last.(IndexSegment)
+		if !ok {
+			return Value{}, pathErrorf(v, "expected an IndexSegment, found %T", last)
+		}
+		if i.int < 0 || i.int > len(v.AsArray()) {
+			return Value{}, pathApplyIndexOutOfBoundsError{found: v.AsArray(), idx: i.int}
+		}
+		// We now want to set set v.AsArray()[i]=to, but we don't want to mutate v. We make
+		// a shallow copy of v and then alter that.
+		cp := make(Array, len(v.AsArray()))
+		copy(cp, v.AsArray())
+		cp[i.int] = to
+		return butLast.Set(src, New(cp)) // Finally, propagate the shallow change back up.
+	case v.IsMap():
+		k, ok := last.(KeySegment)
+		if !ok {
+			return Value{}, pathErrorf(v, "expected a KeySegment, found %T", last)
+		}
+		// We now want to set set v.AsMap()[k]=to, but we don't want to mutate v. We make
+		// a shallow copy of v and then alter that.
+		cp := make(Map, len(v.AsMap()))
+		for k, e := range v.AsMap() {
+			cp[k] = e
+		}
+		cp[k.string] = to                // Assign to the new map
+		return butLast.Set(src, New(cp)) // Finally, propagate the shallow change back up.
+	default:
+		return Value{}, pathErrorf(v, "expected a map or array, found %s", typeString(v))
+	}
+}
+
+// Alter changes the value at p by applying f.
+//
+// To preserve metadata, use [WithGoValue] in conjunction with Alter:
+//
+//	p.Alter(v, func(v Value) Value) {
+//		return property.WithGoValue(v, "new-value")
+//	})
+//
+// This will preserve any secrets or dependencies encoded in `v`.
+func (p Path) Alter(v Value, f func(v Value) Value) (Value, PathApplyFailure) {
+	oldValue, err := p.Get(v)
+	if err != nil {
+		return Value{}, err
+	}
+	return p.Set(v, f(oldValue))
+}
+
+type PathSegment interface {
+	apply(Value) (Value, PathApplyFailure)
+}
+
+// NewSegment creates a new [PathSegment] suitable for use in [Path].
+func NewSegment[T interface{ string | int }](v T) PathSegment {
+	switch v := any(v).(type) {
+	case string:
+		return KeySegment{v}
+	case int:
+		return IndexSegment{v}
+	default:
+		panic("impossible")
+	}
+}
+
+type PathApplyFailure interface {
+	error
+
+	Found() Value
+}
+
+// KeySegment represents a traversal into a map by a key.
+//
+// KeySegment does not support glob ("*") expansion. Values are treated as is.
+type KeySegment struct{ string }
+
+func (k KeySegment) apply(v Value) (Value, PathApplyFailure) {
+	if v.IsMap() {
+		m := v.AsMap()
+		r, ok := m[k.string]
+		if ok {
+			return r, nil
+		}
+		return Value{}, pathApplyKeyMissingError{found: m, needle: k.string}
+	}
+	return Value{}, pathApplyKeyExpectedMapError{found: v}
```

Comment:
do we not prefer negative assertions to reduce nesting?
```suggestion
	if !v.IsMap() {
		return Value{}, pathApplyKeyExpectedMapError{found: v}
	}
	
	m := v.AsMap()
	r, ok := m[k.string]
	if ok {
		return r, nil
	}
	return Value{}, pathApplyKeyMissingError{found: m, needle: k.string}
```

Reply:
(ditto elsewhere)